transport/tlsfrag/fixed_len_writer.go

```diff
@@ -0,0 +1,277 @@
+
+	// the records' sizes and written bytes (including 5 bytes header)
+	r1Size, r1Written, r2Size int
+}
+
+var _ io.Writer = (*recordLenFragWriter)(nil)
+
+// NewRecordLenFuncWriter creates a [io.Writer] that splits the first TLS Client Hello record into two records
+// based on the provided [RecordLenFragFunc] callback.
+// It then writes these records and all subsequent messages to the base [io.Writer].
+// If the first message isn't a Client Hello, no splitting occurs and all messages are written directly to base.
+//
+// The returned [io.Writer] will implement the [io.ReaderFrom] interface for optimized performance if the base
+// [io.Writer] implements [io.ReaderFrom].
+func NewRecordLenFuncWriter(base io.Writer, frag RecordLenFragFunc) (io.Writer, error) {
+	if base == nil {
+		return nil, errors.New("base writer must not be nil")
+	}
+	if frag == nil {
+		return nil, errors.New("fixed length frag callback function must not be nil")
+	}
+	wr := &recordLenFragWriter{
+		base: base,
+		frag: frag,
+		hdr:  make([]byte, 0, recordHeaderLen),
+	}
+	if rf, ok := base.(io.ReaderFrom); ok {
+		return &fixedLenReaderFrom{wr, rf}, nil
+	}
+	return wr, nil
+}
+
+// Write implements io.Writer.Write. It attempts to split the data received in the first one or more Write call(s)
+// into two TLS records if the data corresponds to a TLS Client Hello record without using any additional buffers.
+func (w *recordLenFragWriter) Write(p []byte) (n int, err error) {
+	if !w.done {
+		if w.tlsHdr == nil {
+			// try to fill r.w.hdr of totally 5 bytes
+			hdrLen := len(w.hdr)
+			n = copy(w.hdr[hdrLen:recordHeaderLen], p)
+			p = p[n:]
+			if w.hdr = w.hdr[:hdrLen+n]; len(w.hdr) < recordHeaderLen {
+				return
+			}
+
+			// construct the structured TLS Handshake header object
+			w.tlsHdr, err = newTLSHandshakeRecordHeader(w.hdr)
+			if err != nil || w.tlsHdr.Validate() != nil || w.updateSplitRecordLens() != nil {
+				// invalid TLS header, or invalid split lens, stop splitting
+				w.done = true
+			} else {
+				// update header to be the first record's header
+				w.tlsHdr.SetPayloadLen(uint16(w.r1Size - recordHeaderLen))
+			}
+		}
+
+		if !w.done && w.r1Written < w.r1Size {
+			var m int
+			if w.r1Written < recordHeaderLen {
+				// write the first record's header
+				m, err = w.base.Write(w.tlsHdr[w.r1Written:recordHeaderLen])
+				if w.r1Written += m; w.r1Written < recordHeaderLen || err != nil || len(p) == 0 {
+					return
+				}
+			}
+
+			// write the first record's remaining payload
+			m, err = writeN(w.base, p, w.r1Size-w.r1Written)
+			w.r1Written += m
+			n += m
+			p = p[m:]
+
+			if w.r1Written == w.r1Size {
+				// update w.tlsHdr (aliases w.hdr) to the second record's header
+				w.tlsHdr.SetPayloadLen(uint16(w.r2Size - recordHeaderLen))
+				w.done = true
+			}
+			if w.r1Written < w.r1Size || err != nil {
+				return
+			}
+		}
+	}
+
+	if len(w.hdr) > 0 {
+		m, e := w.base.Write(w.hdr)
```

Comment:
This is doing two writes, so it will split at the TCP layer as well. We may not want that.
Can we split the message without the two writes?

Reply:
The code is still doing multiple writes.

Reply:
Yup, I'm updating the code and not pushed yet.

Reply:
updated